xray/client_test.go

```diff
@@ -97,275 +121,301 @@ func TestRoundTrip(t *testing.T) {
+			assert.False(t, connectSeg.InProgress)
+			assert.True(t, connectSeg.Fault)
+			assert.NotEmpty(t, connectSeg.Subsegments)
+		}
+	}
 }
 
 func TestRoundTripReuseDatarace(t *testing.T) {
-	TestDaemon.Reset()
+	ctx, td := NewTestDaemon()
+	defer td.Close()
+
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		b := []byte(`200 - Nothing to see`)
 		w.WriteHeader(http.StatusOK)
-		w.Write(b)
+		w.Write([]byte(`200 - Nothing to see`))
 	}))
-
 	defer ts.Close()
 
-	wg := sync.WaitGroup{}
-	n := 30
+	client := Client(nil)
+
+	var wg sync.WaitGroup
+	n := 100
 	wg.Add(n)
 	for i := 0; i < n; i++ {
 		go func() {
 			defer wg.Done()
-			reader := strings.NewReader("")
-			ctx, root := BeginSegment(context.Background(), "Test")
-			req, _ := http.NewRequest("GET", strings.Replace(ts.URL, "127.0.0.1", "localhost", -1), reader)
-			req = req.WithContext(ctx)
-			res, err := rt.RoundTrip(req)
-			ioutil.ReadAll(res.Body)
-			res.Body.Close() // make net/http/transport.go connection reuse
-			root.Close(nil)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			err := httpDoTest(ctx, client, http.MethodGet, ts.URL, nil)
 			assert.NoError(t, err)
 		}()
 	}
+	wg.Wait()
+
 	for i := 0; i < n; i++ {
-		_, e := TestDaemon.Recv()
-		assert.NoError(t, e)
+		_, err := td.Recv()
+		assert.NoError(t, err)
 	}
-	wg.Wait()
 }
 
 func TestRoundTripReuseTLSDatarace(t *testing.T) {
-	TestDaemon.Reset()
+	ctx, td := NewTestDaemon()
+	defer td.Close()
+
 	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		b := []byte(`200 - Nothing to see`)
 		w.WriteHeader(http.StatusOK)
-		w.Write(b)
+		if _, err := w.Write([]byte(`200 - Nothing to see`)); err != nil {
+			panic(err)
+		}
 	}))
 	defer ts.Close()
 
-	certpool := x509.NewCertPool()
-	certpool.AddCert(ts.Certificate())
-	tr := &http.Transport{
-		TLSClientConfig: &tls.Config{
-			RootCAs: certpool,
-		},
-	}
-	rt := &roundtripper{
-		Base: tr,
-	}
+	client := Client(ts.Client())
 
-	wg := sync.WaitGroup{}
-	n := 30
+	var wg sync.WaitGroup
+	n := 100
 	wg.Add(n)
 	for i := 0; i < n; i++ {
 		go func() {
 			defer wg.Done()
-			reader := strings.NewReader("")
-			ctx, root := BeginSegment(context.Background(), "Test")
-			req, _ := http.NewRequest("GET", ts.URL, reader)
-			req = req.WithContext(ctx)
-			res, err := rt.RoundTrip(req)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			err := httpDoTest(ctx, client, http.MethodGet, ts.URL, nil)
 			assert.NoError(t, err)
-			ioutil.ReadAll(res.Body)
-			res.Body.Close() // make net/http/transport.go connection reuse
-			root.Close(nil)
 		}()
 	}
+	wg.Wait()
+
 	for i := 0; i < n; i++ {
-		_, e := TestDaemon.Recv()
-		assert.NoError(t, e)
+		_, err := td.Recv()
+		assert.NoError(t, err)
 	}
-	wg.Wait()
 }
 
-func TestRoundTripHttp2Datarace(t *testing.T) {
-	TestDaemon.Reset()
+func TestRoundTripReuseHTTP2Datarace(t *testing.T) {
```

Comment:
would you mind explaining change in this function ?

Reply:
it seems that `http2.ConfigureTransport(tr)` doesn't work we want.
we need to configure the server not only the client.

```go
	// configure http/2
	if err := http2.ConfigureServer(ts.Config, nil); !assert.NoError(t, err) {
		return
	}
	ts.TLS = ts.Config.TLSConfig
	ts.StartTLS()
```

oh.. Here is my mistake of copy.
It's from `TestRoundTripReuseTLSDatarace`

```go
	var wg sync.WaitGroup
	n := 100
	wg.Add(n)
	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()
			ctx, cancel := context.WithCancel(ctx)
			defer cancel()
			err := httpDoTest(ctx, client, http.MethodGet, ts.URL, nil)
			assert.NoError(t, err)
		}()
	}
	wg.Wait()
```